src/compiler/checker.ts

```diff
@@ -7039,12 +7041,28 @@ namespace ts {
                 hasSpreadElement = hasSpreadElement || e.kind === SyntaxKind.SpreadElementExpression;
             }
             if (!hasSpreadElement) {
+                // If array literal is actually a destructuring pattern, mark it as an implied type. We do this such
+                // that we get the same behavior for "var [x, y] = []" and "[x, y] = []".
+                if (inDestructuringPattern && elementTypes.length) {
+                    return createImpliedType(createTupleType(elementTypes));

```

Comment:
If we want to allow `var { x, y } = {}` and the corresponding assignment `{ x, y } = {}` then we'd have to do it for object literals as well.